fggs/indices.py

```diff
@@ -0,0 +1,1293 @@
+indices represented by its "free" EmbeddingVars to a virtual index.  The
+embeddings define an affine transform from the storage offset and strides of
+the virtual tensor (if it ever gets materialized) to the storage offset and
+strides of the physical tensor (a view on the virtual).
+
+In sum, a virtual tensor is represented by
+- a physical tensor,
+- a sequence of "physical" EmbeddingVars (ordered according to the dimensions
+  of the physical tensor), and
+- a sequence of "virtual" embeddings (ordered according to the dimensions of
+  the virtual tensor) containing exactly those EmbeddingVars.
+We store these three pieces of information together in an EmbeddedTensor.
+"""
+
+from __future__ import annotations
+from typing import Sequence, Iterator, List, Tuple, Dict, Set, FrozenSet, Any, Optional, Union, Callable, cast
+from dataclasses import dataclass
+from abc import ABC, abstractmethod
+from warnings import warn
+from itertools import zip_longest, chain, count, product, repeat
+from functools import reduce
+from operator import mul
+from math import inf, log, log1p, exp, expm1, isnan, isinf
+from string import ascii_uppercase
+from sys import float_info#, stderr
+import torch
+from torch import Tensor, Size
+import torch_semiring_einsum
+from fggs.semirings import Semiring
+
+__all__ = ['Embedding', 'EmbeddingVar', 'ProductEmbedding', 'SumEmbedding',
+           'EmbeddedTensor', 'stack', 'einsum', 'log_viterbi_einsum_forward']
+
+Rename = Dict["EmbeddingVar", "EmbeddingVar"]
+
+Subst = Dict["EmbeddingVar", "Embedding"]
+
+AntiSubst = Tuple[Dict[Tuple["Embedding", "Embedding"], "EmbeddingVar"],
+                  Dict["EmbeddingVar", Tuple["Embedding", "Embedding"]]]
+
+NumberType = Union[bool, int, float] # omit the complex type to work around https://github.com/pytorch/pytorch/pull/91345
+
+def make_letterer(alphabet: str = ascii_uppercase) -> Callable[[Any], str]:
+    d : Dict[Any, str] = {}
+    l = (''.join(chars) for n in count(1)
+                        for chars in product(ascii_uppercase, repeat=n))
+    def letter(x: Any) -> str:
+        s = d.get(x)
+        if s is None:
+            s = next(l)
+            d[x] = s
+        return s
+    return letter
+
+debugging_letterer = make_letterer()
+
+class Embedding(ABC):
+    """An injective mapping from physical indices to virtual indices."""
+
+    @abstractmethod
+    def depict(self, letterer: Callable[[EmbeddingVar], str]) -> str:
+        pass
+
+    @abstractmethod
+    def numel(self) -> int:
+        """The virtual indices in the image of this mapping lie in [0,numel())."""
+        pass
+
+    @abstractmethod
+    def fv(self, subst: Subst) -> Iterator[EmbeddingVar]:
+        """Return the set of all EmbeddingVars in self."""
+        pass
+
+    @abstractmethod
+    def stride(self, subst: Subst) -> Tuple[int, Dict[EmbeddingVar, int]]:
+        """The coefficients of the affine map from physical to virtual indices."""
```

Comment:
Please add (if correct): if (offset, strides) = self.stride(), then the physical indices {x1: i1, x2:i2, ...}  map to virtual index offset+strides[x1]+strides[x2]+....